### DISCUSSION TITLE
How to implement an "after success" action?

### DISCUSSION THREAD

**Discussion by GabeDuarteM** [community/community#27141] [Actions]
<p>I have the following action, named “Node CI”, which runs on every push:</p>
<pre><code>name: Node CI

on: [push]

env:
  CI: true

jobs:
  build:

    runs-on: ubuntu-latest

    strategy:
      matrix:
        node-version: [8.x, 13.x]

    steps:
    - uses: actions/checkout@v1
    - name: Use Node.js ${{ matrix.node-version }}
      uses: actions/setup-node@v1
      with:
        node-version: ${{ matrix.node-version }}
    - run: yarn install
    - run: yarn run ci
</code></pre>
<p>After all build matrixes are successful (as in, successful build on both node 8 and 13), if the branch is master, I’d like to run a release process, which would consist of basically</p>
<pre><code>name: Release
on:
  push:
    branches:
      - master
jobs:
  release:
    name: Release
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v1
      - name: Setup Node.js
        uses: actions/setup-node@v1
        with:
          node-version: 13.x
      - name: Install dependencies
        run: yarn install
- name: Build
run: yarn run build
      - name: Release
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
        run: npx semantic-release
</code></pre>
<p>How can I make the “Release” workflow run only after all “Node CI” are finished and successful (including all matrixes)?</p>

**Comment by nakedible** [comment, marked as answer]
<p>Put them both in the same workflow, and add:</p>
<pre><code>jobs:
  build:
    ...
  release:
    if: ${{ github.ref == 'master' }}
    needs: build
    ...
</code></pre>
<p>Release job only gets triggered if we are in master branch and “build” job is successful and complete, including the whole matrix.</p>

**Comment by Multimegamander** [comment]
<p>How would one implement an “else” script, such as using if for success, and something else for failure (what I’m trying to do)</p>

**Comment by benstear** [comment]
<p>Any solution to the if/else problem?</p>

**Comment by jenschelkopf** [comment]
<p>Take a look at the Workflow Conclusion Action.</p>
<p><a href="https://github.com/marketplace/actions/workflow-conclusion-action" rel="nofollow noopener">https://github.com/marketplace/actions/workflow-conclusion-action</a></p>
<p>The action creates an environment variable that you can use check the status of the other jobs.</p>
<p>Here’s the example from their README:</p>
<pre><code>on: push

name: CI

jobs:
  lint:
    name: ESLint
    runs-on: ubuntu-latest
    ...

  test:
    name: Coverage
    needs: lint
    strategy:
      matrix:
        node: ['11', '12']
    ...

  publish:
    name: Publish Package
    needs: test
    if: startsWith(github.ref, 'refs/tags/v')
    ...

  slack:
    name: Slack
    needs: publish # set needs only last job except this job
    runs-on: ubuntu-latest
    if: always() # set always
    steps:
        # run this action to get workflow conclusion
        # You can get conclusion by env (env.WORKFLOW_CONCLUSION)
      - uses: technote-space/workflow-conclusion-action@v1
      - uses: 8398a7/action-slack@v2
        with:
          # status: ${{ env.WORKFLOW_CONCLUSION }} # neutral, success, cancelled, timed_out, failure
          status: failure
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          SLACK_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
        if: env.WORKFLOW_CONCLUSION == 'failure' # notify only if failure
</code></pre>
<p>Notice that the last step only runs if the other jobs failed.</p>

**Comment by brettins** [comment]
<p>This is how I implemented if/else with slack notifications for my failed builds:</p>
<pre><code class="lang-auto">jobs:
  build-and-deploy:
    name: Build and Deploy
    runs-on: ubuntu-latest
    steps:
    ...
   - name: notify of failure
      if: ${{ failure() }}
      uses: tokorom/action-slack-incoming-webhook@main
      env:
        INCOMING_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
      with:
        text: XXX Build failed - Deploy - Server - TEST - Azure Web Apps CI/CD. XXX
    - name: notify of success
      if: ${{ success() }}
     uses: tokorom/action-slack-incoming-webhook@main
      env:
        INCOMING_WEBHOOK_URL: ${{ secrets.SLACK_WEBHOOK_URL }}
      with:
        text: ":heavy_check_mark: Build success - Deploy -  Server - TEST - Azure Web pps CI/CD. :heavy_check_mark:"
</code></pre>

**Comment by case** [comment]
<p>This is a great solution. You can also do it with <code>curl</code>:</p>
<pre><code class="lang-auto">      - name: Notify if failure
        if: ${{ failure() }}
        run: |
          curl -X POST -H 'Content-type: application/json' --data '{"text":"❌ {your-job-name-here} failed."}' ${{ secrets.SLACK_WEBHOOK_URL }}

      - name: Notify if success
        if: ${{ success() }}
        run: |
          curl -X POST -H 'Content-type: application/json' --data '{"text":"✅ {your-job-name-here} passed."}' ${{ secrets.SLACK_WEBHOOK_URL }}
</code></pre>

**Reply by ridwaanzan** [reply]
Thanks, this is what i need! i'll try it.
